### PR TITLE
FIXME: Set correct locus for INSERT RETURNING in a CTE

### PR DESCRIPTION
**Pull request by yaowangm**
### Brief

For statement of INSERT RETURNING in a CTE, existing code always sets locus to Strewn, which causes that an extra Redistribute Motion is always needed even when the distribution of the CTE is compatible to upper node. The fix is to set correct locus according to the distribution of RETURNING of the CTE, so extra Redistribute Motion may be avoided for the case.

### Background

It's from a FIXME at /src/backend/optimizer/util/pathnode.c:5408:
```
    /*   
     * Set the distribution of the ModifyTable node itself. If there is only
     * one subplan, or all the subplans have a compatible distribution, then
     * we could mark the ModifyTable with the same distribution key. However,
     * currently, because a ModifyTable node can only be at the top of the
     * plan, it won't make any difference to the overall plan.
     *
     * GPDB_96_MERGE_FIXME: it might with e.g. a INSERT RETURNING in a CTE
     * I tried here, the locus setting is quite simple, but failed if it's not
     * in a CTE and the locus is General. Haven't figured out how to create
     * flow in that case.
     * Example:
     * CREATE TABLE cte_returning_locus(c1 int) DISTRIBUTED BY (c1);
     * COPY cte_returning_locus FROM PROGRAM 'seq 1 100';
     * EXPLAIN WITH aa AS (
     *        INSERT INTO cte_returning_locus SELECT generate_series(3,300) RETURNING c1
     * )
     * SELECT count(*) FROM aa,cte_returning_locus WHERE aa.c1 = cte_returning_locus.c1;
     *
     * The returning doesn't need a motion to be hash joined, works fine. But
     * without the WITH, what is the proper flow? FLOW_SINGLETON returns
     * nothing, FLOW_PARTITIONED without hashExprs(General locus has no
     * distkeys) returns duplication.
     *
     * GPDB_90_MERGE_FIXME: I've hacked a basic implementation of the above for
     * the case where all the subplans are POLICYTYPE_ENTRY, but it seems like
     * there should be a more general way to do this.
     */
```
And the issue actually comes from two separated commits:
https://github.com/greenplum-db/gpdb/pull/8955/commits/4e712e3fdc7c8e1d671a80217a85f028597a39dc by Adam
https://github.com/greenplum-db/gpdb/pull/8955/commits/2a864ff1421a631be1ebd2fb68f93cdfb8df09a4 by Heikki

### Example

Script:
```
CREATE TABLE cte_returning_locus(c1 int) DISTRIBUTED BY (c1);
COPY cte_returning_locus FROM PROGRAM 'seq 1 100';
EXPLAIN WITH aa AS (
  INSERT INTO cte_returning_locus SELECT generate_series(3,300)
  RETURNING c1)
SELECT count(*) FROM aa,cte_returning_locus
  WHERE aa.c1 = cte_returning_locus.c1;
```
The plan for existing code:
```
testdb=# explain verbose WITH aa AS (
testdb(# INSERT INTO cte_returning_locus SELECT generate_series(3,300) RETURNING c1
testdb(# )
testdb-# SELECT count(*) FROM aa,cte_returning_locus WHERE aa.c1 = cte_returning_locus.c1;
                                                            QUERY PLAN-------------------------------------------------------------------------------------------------------------
----------------------
 Finalize Aggregate  (cost=18.96..18.97 rows=1 width=8)
   Output: count(*)
   ->  Gather Motion 3:1  (slice1; segments: 3)  (cost=18.90..18.95 rows=3 width=8)
         Output: (PARTIAL count(*))
         ->  Partial Aggregate  (cost=18.90..18.91 rows=1 width=8)
               Output: PARTIAL count(*)
               ->  Hash Join  (cost=11.69..18.41 rows=199 width=0)
                     Hash Cond: (cte_returning_locus.c1 = cte_returning_locus_1.c1)
                     ->  Seq Scan on public.cte_returning_locus  (cost=0.00..3.99 rows=199 width=4)
                           Output: cte_returning_locus.c1
                     ->  Hash  (cost=10.45..10.45 rows=99 width=4)
                           Output: cte_returning_locus_1.c1
                           ->  Redistribute Motion 3:3  (slice2; segments: 3)  (cost=0.00..10.45 rows=99 widt
h=4)
                                 Output: cte_returning_locus_1.c1
                                 Hash Key: cte_returning_locus_1.c1
                                 ->  Insert on public.cte_returning_locus cte_returning_locus_1  (cost=0.00..
8.46 rows=99 width=4)
                                       Output: cte_returning_locus_1.c1
                                       ->  Redistribute Motion 1:3  (slice3; segments: 1)  (cost=0.00..8.46 r
ows=99 width=4)
                                             Output: (generate_series(3, 300))
                                             Hash Key: (generate_series(3, 300))
                                             ->  ProjectSet  (cost=0.00..1.51 rows=298 width=4)
                                                   Output: generate_series(3, 300)
                                                   ->  Result  (cost=0.00..0.01 rows=1 width=0)
 Optimizer: Postgres-based planner
 Settings: optimizer = 'off'
(25 rows) 
```
We can see there is Redistribute Motion (3:3) in front of the Insert node.

With the fix:
```
testdb=# EXPLAIN verbose WITH aa AS (
testdb(#   INSERT INTO cte_returning_locus SELECT generate_series(3,300)
testdb(#   RETURNING c1)
testdb-# SELECT count(*) FROM aa,cte_returning_locus
testdb-#   WHERE aa.c1 = cte_returning_locus.c1;
                                                         QUERY PLAN

-----------------------------------------------------------------------------------------
------------------------------------
 Finalize Aggregate  (cost=16.83..16.84 rows=1 width=8)
   Output: count(*)
   ->  Gather Motion 3:1  (slice1; segments: 3)  (cost=16.77..16.82 rows=3 width=8)
         Output: (PARTIAL count(*))
         ->  Partial Aggregate  (cost=16.77..16.78 rows=1 width=8)
               Output: PARTIAL count(*)
               ->  Hash Join  (cost=9.70..16.20 rows=230 width=0)
                     Hash Cond: (cte_returning_locus.c1 = cte_returning_locus_1.c1)
                     ->  Seq Scan on public.cte_returning_locus  (cost=0.00..3.32 rows=23
2 width=4)
                           Output: cte_returning_locus.c1
                     ->  Hash  (cost=8.46..8.46 rows=99 width=4)
                           Output: cte_returning_locus_1.c1
                           ->  Insert on public.cte_returning_locus cte_returning_locus_1
  (cost=0.00..8.46 rows=99 width=4)
                                 Output: cte_returning_locus_1.c1
                                 ->  Redistribute Motion 1:3  (slice2; segments: 1)  (cos
t=0.00..8.46 rows=99 width=4)
                                       Output: (generate_series(3, 300))
                                       Hash Key: (generate_series(3, 300))
                                       ->  ProjectSet  (cost=0.00..1.51 rows=298 width=4)
                                             Output: generate_series(3, 300)
                                             ->  Result  (cost=0.00..0.01 rows=1 width=0)
 Optimizer: Postgres-based planner
(21 rows)
```
The extra Redistribute Motion is gone because of the correct locus.

### Notes

Please note the assumption that for a INSERT there must be only one sub path (and one item in returningLists). I didn't find any negative example, but I am not very sure about the assumption. Please help to confirm it. Thanks!

### CODE REVIEW

**Comment by kainwen** [comment]
I don't think we need to do anything on this part of code. Current code is simple, correct and it works.

**Comment by avamingli** [comment]
It's not enough  only to with returning, this issue has been (partly or fully) resolved in https://github.com/greenplum-db/gpdb/pull/15862, and let insert with returning join with others.

**Comment by yaowangm** [comment]
After discussion we decided it's not worthy to do because the scenario is not common and the benefit is limited. I will close this PR and create another PR (https://github.com/greenplum-db/gpdb/pull/16980) to simply remove the FIXME with adding some comments.

File: src/backend/optimizer/util/pathnode.c [hunk 1]
```diff
@@ -5280,6 +5281,31 @@ create_modifytable_path(PlannerInfo *root, RelOptInfo *rel,
...
+	 *   WHERE aa.c1 = cte_returning_locus.c1;
+	 *
+	 * For the case, we need to set proper pathtarget according to the 
+	 * returningLists, and the pathtarget will be used by
+	 * cdbpathlocus_from_subquery() later (see cdb_pull_up_eclass() for
+	 * details).
+	 * We assume returningLists must have only one item because there is
+	 * only one subpath for CMD_INSERT.
+	 */
+	if (operation == CMD_INSERT && returningLists)
```
Comment (adam8157) [inline src/backend/optimizer/util/pathnode.c:5302]: This PR could only handle the INSERT writable CTE?

Comment (yaowangm) [reply src/backend/optimizer/util/pathnode.c:5302]: Yes. Due to the complexity of the scenarios, I planned to support other possible kinds via separated PR.

**Review by adam8157 — COMMENTED** [review]
The scenario is not common and it's not worth it either, but I will +1 if someone fixed not only INSERT but also other writable CTEs.

**Comment by yaowangm** [comment]
> It's not enough only to with returning, this issue has been (partly or fully) resolved in #15862, and let insert with returning join with others.

@avamingli I took a basic look and prefer #15862 is not directly related to the issue I handled. I will try to find time to review the PR. Thanks!

**Pull request closed without merging**